cluster/azure: kube-up works; validation succeeds

colemickens opened the pull request.

colemickens committed:
cluster/azure: kube-up works; validation succeeds

colemickens commented:
I'm also working on E2E Jenkins Job definitions that builds on top of this.

davidopp commented on cluster/azure/config-default.sh:
```diff
@@ -0,0 +1,31 @@
+# cp config-default.sh config-real.sh
+# export KUBE_CONFIG_FILE=`pwd`/config-real.sh
+
+AZURE_LOCATION="westus" # this must be in the FQDN format so that url construction works in util.sh
```
I'm confused about the comment; this isn't a FQDN, it's just a hostname

davidopp commented:
Style comment: Can you please move all the shell comments to the line above rather than at the end fo the line?

davidopp commented on cluster/azure/util.sh:
```diff
@@ -0,0 +1,167 @@
+    date_start="$(date)"
+    echo "++> AZURE KUBE-UP STARTED: ${date_start}"
+
+    export AZURE_HYPERKUBE_SPEC="gcr.io/google_containers/hyperkube:v1.1.8"
```
I'm not 100% sure but I think maybe the right way to do this is
gcr.io/google_containers/hyperkube-amd64:v${K8S_VERSION}

In any event, something parameterized seems better than hard-coded.

davidopp commented:
Thanks a lot for doing this. I just had a few minor comments.

colemickens committed:
respond to davidopp's feedback

also:
- adds a mechanism to build and upload hyperkube for non-official
  releases
- adds a mechanism for proxying azkube's traffic
- --no-cloud-provider for now
- support specifying the resource group for CI scenarios

colemickens commented:
@davidopp Thanks for the feedback. I'm in progress on parameterizing for the version but have run into a couple of things:

1. The cluster is not coming up correctly. I'm getting a x509 error after switching from `v1.1.8` to latest `master`. I'm digging into that now.

2. It seems the `hyperkube` docker image is not always built as part of the release. It looks like it's only built when some environment variables are set. The changes I'm making need the hyperkube image to be built as part of `make release` so that the script can upload it to the user-specified repo for when they're not using official builds. (relevant: https://github.com/kubernetes/kubernetes/pull/21928 and https://github.com/kubernetes/kubernetes/pull/20710)

davidopp commented:
@fgrzadkowski or @david-mcmahon can you comment on (2) from above (viz. https://github.com/kubernetes/kubernetes/pull/21928#issuecomment-188828940 ) ?

colemickens commented:
Another issue after moving to `master`: Cluster bring-up is failing because the server times out:

```
../../_output/dockerized/bin/linux/amd64/kubectl --v=10 get cs
I0302 15:49:16.811006   31111 loader.go:229] Config loaded from file /home/cole/.kube/config
I0302 15:49:17.395050   31111 round_trippers.go:267] curl -k -v -XGET  -H "Accept: application/json, */*" https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api
I0302 15:49:17.973167   31111 round_trippers.go:286] GET https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api 200 OK in 578 milliseconds
I0302 15:49:17.973183   31111 round_trippers.go:292] Response Headers:
I0302 15:49:17.973198   31111 round_trippers.go:295]     Date: Wed, 02 Mar 2016 23:49:17 GMT
I0302 15:49:17.973205   31111 round_trippers.go:295]     Content-Length: 130
I0302 15:49:17.973211   31111 round_trippers.go:295]     Content-Type: application/json
I0302 15:49:17.973254   31111 request.go:870] Response Body: {"kind":"APIVersions","versions":["v1"],"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.0.1.4:443"}]}
I0302 15:49:17.973433   31111 round_trippers.go:267] curl -k -v -XGET  -H "Accept: application/json, */*" https://kube-20160302-145225.westus.cloudapp.azure.com:6443/apis
I0302 15:49:18.002468   31111 round_trippers.go:286] GET https://kube-20160302-145225.westus.cloudapp.azure.com:6443/apis 200 OK in 29 milliseconds
I0302 15:49:18.002480   31111 round_trippers.go:292] Response Headers:
I0302 15:49:18.002494   31111 round_trippers.go:295]     Content-Type: application/json
I0302 15:49:18.002506   31111 round_trippers.go:295]     Date: Wed, 02 Mar 2016 23:49:17 GMT
I0302 15:49:18.002512   31111 round_trippers.go:295]     Content-Length: 760
I0302 15:49:18.002546   31111 request.go:870] Response Body: {"kind":"APIGroupList","groups":[{"name":"autoscaling","versions":[{"groupVersion":"autoscaling/v1","version":"v1"}],"preferredVersion":{"groupVersion":"autoscaling/v1","version":"v1"},"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.0.1.4:443"}]},{"name":"batch","versions":[{"groupVersion":"batch/v1","version":"v1"}],"preferredVersion":{"groupVersion":"batch/v1","version":"v1"},"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.0.1.4:443"}]},{"name":"extensions","versions":[{"groupVersion":"extensions/v1beta1","version":"v1beta1"}],"preferredVersion":{"groupVersion":"extensions/v1beta1","version":"v1beta1"},"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.0.1.4:443"}]}]}
I0302 15:49:18.004206   31111 round_trippers.go:267] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.2.0 (linux/amd64) kubernetes/19fde13" https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api/v1/componentstatuses
I0302 15:50:18.051768   31111 round_trippers.go:286] GET https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api/v1/componentstatuses 504 Gateway Timeout in 60047 milliseconds
I0302 15:50:18.051792   31111 round_trippers.go:292] Response Headers:
I0302 15:50:18.051806   31111 round_trippers.go:295]     Date: Wed, 02 Mar 2016 23:50:17 GMT
I0302 15:50:18.051810   31111 round_trippers.go:295]     Content-Length: 189
I0302 15:50:18.051815   31111 round_trippers.go:295]     Content-Type: text/plain; charset=utf-8
I0302 15:50:18.051856   31111 request.go:870] Response Body: {"ErrStatus":{"metadata":{},"status":"Failure","message":"The  operation against  could not be completed at this time, please try again.","reason":"ServerTimeout","details":{},"code":500}}
I0302 15:50:18.051959   31111 request.go:933] Response Body: {"ErrStatus":{"metadata":{},"status":"Failure","message":"The  operation against  could not be completed at this time, please try again.","reason":"ServerTimeout","details":{},"code":500}}
I0302 15:50:18.052093   31111 helpers.go:172] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server cannot complete the requested operation at this time, try again later (get componentstatuses)",
  "reason": "ServerTimeout",
  "details": {
    "kind": "componentstatuses",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "{\"ErrStatus\":{\"metadata\":{},\"status\":\"Failure\",\"message\":\"The  operation against  could not be completed at this time, please try again.\",\"reason\":\"ServerTimeout\",\"details\":{},\"code\":500}}"
      }
    ]
  },
  "code": 504
}]
F0302 15:50:18.052112   31111 helpers.go:107] Error from server: the server cannot complete the requested operation at this time, try again later (get componentstatuses)
```

An interesting note here; it might be relevant. When I query with `kubectl` specifying the auth as command line flags, rather than letting kubectl using `~/.kube/config`, I can see it trying to hit the namespace'd version of the componentstatuses endpoint. And it results in a 404:

```
$ kubectl --cluster=kube-20160302-145225 --context=kube-20160302-145225 --client-certificate=/home/cole/code/azure/kubernetes_gopath/src/github.com/kubernetes/kubernetes/cluster/azure/_deployments/kube-20160302-145225/client.crt --client-key=/home/cole/code/azure/kubernetes_gopath/src/github.com/kubernetes/kubernetes/cluster/azure/_deployments/kube-20160302-145225/client.key --certificate-authority=/home/cole/code/azure/kubernetes_gopath/src/github.com/kubernetes/kubernetes/cluster/azure/_deployments/kube-20160302-145225/ca.crt --server=https://kube-20160302-145225.westus.cloudapp.azure.com:6443 --v=10 get cs

I0302 15:44:56.067418   27683 debugging.go:101] curl -k -v -XGET  -H "User-Agent: kubectl/v1.0.3 (linux/amd64) kubernetes/61c6ac5" https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api
I0302 15:44:56.485476   27683 debugging.go:120] GET https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api 200 OK in 418 milliseconds
I0302 15:44:56.485491   27683 debugging.go:126] Response Headers:
I0302 15:44:56.485505   27683 debugging.go:129]     Date: Wed, 02 Mar 2016 23:44:56 GMT
I0302 15:44:56.485518   27683 debugging.go:129]     Content-Length: 130
I0302 15:44:56.485522   27683 debugging.go:129]     Content-Type: application/json
I0302 15:44:56.485551   27683 request.go:779] Response Body: {"kind":"APIVersions","versions":["v1"],"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.0.1.4:443"}]}
I0302 15:44:57.067666   27683 debugging.go:101] curl -k -v -XGET  -H "User-Agent: kubectl/v1.0.3 (linux/amd64) kubernetes/61c6ac5" https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api/v1/namespaces/default/componentstatuses
I0302 15:44:57.453953   27683 debugging.go:120] GET https://kube-20160302-145225.westus.cloudapp.azure.com:6443/api/v1/namespaces/default/componentstatuses 404 Not Found in 386 milliseconds
I0302 15:44:57.453976   27683 debugging.go:126] Response Headers:
I0302 15:44:57.453989   27683 debugging.go:129]     Content-Type: application/json
I0302 15:44:57.453993   27683 debugging.go:129]     Date: Wed, 02 Mar 2016 23:44:57 GMT
I0302 15:44:57.453997   27683 debugging.go:129]     Content-Length: 174
I0302 15:44:57.454042   27683 request.go:779] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
I0302 15:44:57.454629   27683 selector.go:53] Unable to list "componentstatuses": the server could not find the requested resource
```

colemickens commented:
(With a more up-to-date kubectl, it doesn't try to hit the namespace'd cs endpoint, but it is still timing out)

colemickens commented:
Looks like it's timing out when querying one of the backends, even though I'm able to hit the scheduler, replication-controller and etcd (from the host at least) on their health check endpoints and they all report green.

If there are no other suggestions soon I might add some additional logging and try with my own builds.

Here's the failure from the apiserver's perspective, with `--v=100`:

```
I0303 01:25:30.251694       1 handlers.go:152] GET /api/v1/componentstatuses: (1m0.000300943s) 504
goroutine 1122 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc208b36230, 0x1f8)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc208b36230, 0x1f8)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
k8s.io/kubernetes/pkg/apiserver.(*baseTimeoutWriter).timeout(0xc20898b620, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:265 +0xbf
k8s.io/kubernetes/pkg/apiserver.(*timeoutHandler).ServeHTTP(0xc208cf6c00, 0x7f9515187cd0, 0xc208b36230, 0xc20a12e270)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:189 +0x355
k8s.io/kubernetes/pkg/apiserver.func·014(0x7f9515187c98, 0xc20983c1e0, 0xc20a12e270)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:151 +0x1cb
net/http.HandlerFunc.ServeHTTP(0xc208d20e60, 0x7f9515187c98, 0xc20983c1e0, 0xc20a12e270)
        /usr/src/go/src/net/http/server.go:1265 +0x41
k8s.io/kubernetes/pkg/apiserver.func·012(0x7f9515187c98, 0xc20983c1e0, 0xc20a12e270)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:111 +0x171
net/http.HandlerFunc.ServeHTTP(0xc208cf6c20, 0x7f9515187c98, 0xc20983c1e0, 0xc20a12e270)
        /usr/src/go/src/net/http/server.go:1265 +0x41
net/http.serverHandler.ServeHTTP(0xc20888ce40, 0x7f9515187c98, 0xc20983c1e0, 0xc20a12e270)
        /usr/src/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc208f3a1e0)
        /usr/src/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
        /usr/src/go/src/net/http/server.go:1751 +0x35e
 [[kubectl/v1.2.0 (linux/amd64) kubernetes/19fde13] 66.235.2.123:55078]
E0303 01:25:30.254130       1 apiserver.go:442] apiserver was unable to write a JSON response: http: Handler timeout
E0303 01:25:30.254196       1 errors.go:63] apiserver received an error that is not an unversioned.Status: http: Handler timeout
I0303 01:25:30.254335       1 trace.go:57] Trace "List /api/v1/componentstatuses" (started 2016-03-03 01:24:30.252032476 +0000 UTC):
[5.9µs] [5.9µs] About to List from storage
[1m0.001676839s] [1m0.001670939s] Listing from storage done
[1m0.001800439s] [123.6µs] Self-linking done
[1m0.002265039s] [464.6µs] Writing http response done (5 items)
[1m0.002266439s] [1.4µs] END
```

colemickens commented:
Update: I coded around the need for Hyperkube to be rebuilt. It will now detect if hyperkube hasn't been built for the current KUBE_GIT_VERSION and will build, tag, and upload it to the user specified repo.

Still looking into why apiserver is timing out on the componentstatuses...

fgrzadkowski commented:
`hyperkube` is build during an [official release](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/releasing.md). It's not build when you just type `make release` or `make quick-release` as it would slow down every day iterations when are iterating over a local change.

davidopp commented:
@lavalamp is there a chance you could help @colemickens with the problem he describes here https://github.com/kubernetes/kubernetes/pull/22231#issuecomment-191527409? It would be great to get Azure support into 1.2.

lavalamp commented:
@colemickens the list of components comes from this. https://github.com/kubernetes/kubernetes/blob/43792754d89feb80edd846ba3b40297f2a105e14/pkg/master/master.go#L472

When you hit the components, are you using the same address/ports that are added to that list? One of those addresses is not responding. (I'm sorry that's hardcoded like that...)

Clearly we need to add a link-layer timeout (to the http.client for these checks), that's why you get the stack trace, because the request timeout triggers. Unfortunately we leak a goroutine when this happens... I'll file an issue.

colemickens commented:
@lavalamp I should've sent an update. I figured out what the problem was after adding some trace logging in the location you just pointed at. I had etcd advertising an address that wasn't resolveable from inside the hyperkube container.

I'm not really sure why this was working in v1.1.8 though and am looking into that briefly. I'll update this PR today with a working revision whenever I understand why it worked before. (Should it be trying to health check every address that etcd advertises?)

lavalamp commented:
The loop is there in case you're e.g. running a separate etcd for events,
so it will check every backend of every etcd. I don't think it needs to
check every address etcd opens, we don't even know what control port it's
using.

On Fri, Mar 4, 2016 at 10:46 AM, Cole Mickens <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> I should've sent an update. I
> figured out what the problem was after adding some trace logging in the
> location you just pointed at. I had etcd advertising an address that wasn't
> resolveable from inside the hyperkube container.
>
> I'm not really sure why this was working in v1.1.8 though and am looking
> into that briefly. I'll update this PR today with a working revision
> whenever I understand why it worked before. (Should it be trying to health
> check every address that etcd advertises?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22231#issuecomment-192405181>
> .
>

colemickens commented:
@lavalamp: I'm just trying to understand if the behavior of the etcd health checks changed intentionally or if I've stumbled across a bug (or unintentional change in behavior).

In `v1.1.8`, there is only one etcd backend that is checked (`127.0.0.1`). In `master`, there are backends for both `127.0.0.1` and the host's private ip address (`10.0.1.4`).

(Similarly, querying componentstatuses in `v1.1.8` returns only `etcd-0`, whereas on `master` it returns `etcd-0` and `etcd-1` even though they're the same instance)

Is this expected?

lavalamp commented:
etcd-0 and etcd-1 shouldn't be the same instance? The most probable thing
that changed is that we started shuffling events to a second etcd cluster.

On Fri, Mar 4, 2016 at 3:47 PM, Cole Mickens <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp>: I'm just trying to understand if
> the behavior of the etcd health checks changed intentionally or if I've
> stumbled across a bug (or unintentional change in behavior).
>
> In v1.1.8, there is only one etcd backend that is checked (127.0.0.1). In
> master, there are backends for both 127.0.0.1 and the host's private ip
> address (10.0.1.4).
>
> (Similarly, querying componentstatuses in v1.1.8 returns only etcd-0,
> whereas on master it returns etcd-0 and etcd-1 even though they're the
> same instance)
>
> Is this expected?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22231#issuecomment-192526654>
> .
>

davidopp commented:
cc/ @wojtek-t

davidopp commented:
@colemickens let me know when your PR is updated, and I will review it

colemickens commented:
@davidopp I just pushed the update.

davidopp commented on cluster/azure/util.sh:
```diff
@@ -0,0 +1,248 @@
+        return 0
+    fi
+
+    # should these steps tell them to just immediately tag it with the final user-specified repo?
```
probably should prefix with TODO: but fine to do that in a follow-up PR

davidopp commented on cluster/azure/util.sh:
```diff
@@ -0,0 +1,248 @@
+        # this will export AZURE_HYPERKUBE_SPEC based on whether an official image was found
+        # or if it was uploaded to the user specified docker repository.
+        if ! ensure-hyperkube; then
+            echo "Failed to ensure hyperkube was available. Exitting."
```
s/Exitting/Exiting/
(also fine to fix in a followup)

davidopp commented:
LGTM

Thanks a lot for doing this!

If you can update the docs
https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/azure.md
that would be great. Note the info recently sent to the kubernetes-dev mailing list about docs moving.

I guess we should keep a link to the CoreOS-Azure docs for now, but we should figure out what to do with them long-term. Definitely I think we want to get rid of the node.js part...

cc/ @errordeveloper

errordeveloper commented:
> Definitely I think we want to get rid of the node.js part...

@davidopp @colemickens the way it was done was suboptimal, and Azure Resource Manager (ARM) was not available at the time... It was done using JavaScript as that's the language of the Azure CLI, so it made sense, and something more then shell was clearly needed to manipulate YAML files cleanly. Ideally, if I had time, I would redo it using ARM. At the same time, it'd probably benefit from a few new features in CoreOS also. On the other hand, we are working on Weave CNI plugin, and bottom-up guide like that won't be required as such, so let's discuss if moving it to contrib repo may be a better course of action.

davidopp commented:
@errordeveloper Thanks for your comments. I'll open a new issue for us to figure out what to do.
